nexus/pkg/artifacts/downloader.go

```diff
@@ -0,0 +1,309 @@
+package artifacts
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/Khan/genqlient/graphql"
+	"github.com/wandb/wandb/nexus/internal/filetransfer"
+	"github.com/wandb/wandb/nexus/internal/gql"
+	"github.com/wandb/wandb/nexus/pkg/env"
+)
+
+const BATCH_SIZE int = 10000
+const MAX_BACKLOG int = 10000
+
+type ArtifactDownloader struct {
+	// Resources
+	Ctx             context.Context
+	GraphqlClient   graphql.Client
+	DownloadManager *filetransfer.FileTransferManager
+	// Input
+	QualifiedName          string
+	DownloadRoot           *string
+	Recursive              *bool
+	AllowMissingReferences *bool
+	// Properties
+	UpstreamArtifacts   []gql.ArtifactByIDArtifact
+	DefaultDownloadRoot *string
+}
+
+func NewArtifactDownloader(
+	ctx context.Context,
+	graphQLClient graphql.Client,
+	downloadManager *filetransfer.FileTransferManager,
+	qualifiedName string,
+	downloadRoot *string,
+	recursive *bool,
+	allowMissingReferences *bool,
+) *ArtifactDownloader {
+	return &ArtifactDownloader{
+		Ctx:                    ctx,
+		GraphqlClient:          graphQLClient,
+		DownloadManager:        downloadManager,
+		QualifiedName:          qualifiedName,
+		DownloadRoot:           downloadRoot,
+		Recursive:              recursive,
+		AllowMissingReferences: allowMissingReferences,
+		UpstreamArtifacts:      []gql.ArtifactByIDArtifact{},
+		DefaultDownloadRoot:    nil,
+	}
+}
+
+func (ad *ArtifactDownloader) getArtifact() (attrs gql.ArtifactByNameProjectArtifact, rerr error) {
+	entityName, projectName, artifactName, err := parseArtifactQualifiedName(ad.QualifiedName)
+	if err != nil {
+		return gql.ArtifactByNameProjectArtifact{}, err
+	}
+	response, err := gql.ArtifactByName(
+		ad.Ctx,
+		ad.GraphqlClient,
+		entityName,
+		projectName,
+		artifactName,
+	)
+	if err != nil {
+		return gql.ArtifactByNameProjectArtifact{}, err
+	}
+	return *response.GetProject().GetArtifact(), nil
+}
+
+func loadManifestFromURL(url string) (Manifest, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return Manifest{}, err
+	}
+	defer resp.Body.Close()
+	manifest := Manifest{}
+	if resp.StatusCode == http.StatusOK {
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return Manifest{}, fmt.Errorf("error reading response body: %v", err)
+		}
+		err = json.Unmarshal(body, &manifest)
+		if err != nil {
+			return Manifest{}, nil
+		}
+	} else {
+		return Manifest{}, fmt.Errorf("request to get manifest from url failed with status code: %d", resp.StatusCode)
+	}
```

Comment:
Nit: prefer early return rather than indentation.
```
	if resp.StatusCode != http.StatusOK {
		return Manifest{}, fmt.Errorf("request to get manifest from url failed with status code: %d", resp.StatusCode)
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return Manifest{}, fmt.Errorf("error reading response body: %v", err)
	}
	err = json.Unmarshal(body, &manifest)
	if err != nil {
		return Manifest{}, nil
	}
```